server/src/main/java/io/druid/server/coordinator/SegmentProcessingThrottler.java

```diff
@@ -28,107 +28,94 @@
   {
+    this.type = type;
     updateParams(maxReplicants, maxLifetime);
   }
 
   public void updateParams(int maxReplicants, int maxLifetime)
   {
     this.maxReplicants = maxReplicants;
     this.maxLifetime = maxLifetime;
+    this.enableThrottling = maxReplicants > 0;
   }
 
-  public void updateReplicationState(String tier)
+  public void updateState(String tier)
   {
-    update(tier, currentlyReplicating, replicatingLookup, "create");
-  }
-
-  public void updateTerminationState(String tier)
-  {
-    update(tier, currentlyTerminating, terminatingLookup, "terminate");
-  }
-
-  private void update(String tier, ReplicatorSegmentHolder holder, Map<String, Boolean> lookup, String type)
-  {
-    int size = holder.getNumProcessing(tier);
+    int size = currentlyProcessing.getNumProcessing(tier);
     if (size != 0) {
       log.info(
-          "[%s]: Replicant %s queue still has %d segments. Lifetime[%d]. Segments %s",
+          "[%s]: Processing %s queue still has %d segments. Lifetime[%d]. Segments %s",
           tier,
           type,
           size,
-          holder.getLifetime(tier),
-          holder.getCurrentlyProcessingSegmentsAndHosts(tier)
+          currentlyProcessing.getLifetime(tier),
+          currentlyProcessing.getCurrentlyProcessingSegmentsAndHosts(tier)
       );
-      holder.reduceLifetime(tier);
-      lookup.put(tier, false);
+      currentlyProcessing.reduceLifetime(tier);
+      processingLookup.put(tier, false);
 
-      if (holder.getLifetime(tier) < 0) {
-        log.makeAlert("[%s]: Replicant %s queue stuck after %d+ runs!", tier, type, maxLifetime)
-           .addData("segments", holder.getCurrentlyProcessingSegmentsAndHosts(tier))
+      if (currentlyProcessing.getLifetime(tier) < 0) {
+        log.makeAlert("[%s]: Processing %s queue stuck after %d+ runs!", tier, type, maxLifetime)
+           .addData("segments", currentlyProcessing.getCurrentlyProcessingSegmentsAndHosts(tier))
            .emit();
       }
     } else {
-      log.info("[%s]: Replicant %s queue is empty.", tier, type);
-      lookup.put(tier, true);
-      holder.resetLifetime(tier);
+      log.info("[%s]: Processing %s queue is empty.", tier, type);
+      processingLookup.put(tier, true);
+      currentlyProcessing.resetLifetime(tier);
     }
   }
 
-  public boolean canCreateReplicant(String tier)
-  {
-    return replicatingLookup.get(tier) && !currentlyReplicating.isAtMaxReplicants(tier);
-  }
-
-  public boolean canDestroyReplicant(String tier)
-  {
-    return terminatingLookup.get(tier) && !currentlyTerminating.isAtMaxReplicants(tier);
-  }
-
-  public void registerReplicantCreation(String tier, String segmentId, String serverId)
+  public boolean canProcessingSegment(String tier)
   {
-    currentlyReplicating.addSegment(tier, segmentId, serverId);
+    return !enableThrottling || (processingLookup.get(tier) && !currentlyProcessing.isAtMaxReplicants(tier));
```

Comment:
nice catch, yes it can be null in case someone adds a tier after the state of throttler is updated.